SCons/Util/stats.py

```diff
@@ -0,0 +1,215 @@
+# MIT License
+#
+# Copyright The SCons Foundation
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
+# KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+"""
+This package provides a way to gather various statistics during a SCons run and dump that info in several formats
+
+Additionally, it probably makes sense to do stderr/stdout output of those statistics here as well
+
+There are basically two types of stats.
+1. Timer (start/stop/time) for specific event.  These events can be hierarchical. So you can record the children events of some parent.
+   Think program compile could contain the total Program builder time, which could include linking, and stripping the executable
+2. Counter. Counting the number of events and/or objects created. This would likely only be reported at the end of a given SCons run,
+   though it might be useful to query during a run.
+
+"""
+from abc import ABC
+
+import platform
+import json
+import sys
+from datetime import datetime
+
+import SCons.Debug
+
+all_stats = {}
+ENABLE_JSON = False
+JSON_OUTPUT_FILE = 'scons_stats.json'
+
+def add_stat_type(name, stat_object):
+    """
+    Add a statistic type to the global collection
+    """
+    if name in all_stats:
+        raise UserWarning(f'Stat type {name} already exists')
+    else:
+        all_stats[name] = stat_object
+
+
+class Stats(ABC):
+    def __init__(self):
+        self.stats = []
+        self.labels = []
+        self.append = self.do_nothing
+        self.print_stats = self.do_nothing
+        self.enabled = False
+
+    def do_append(self, label):
```

Comment:
you can leave the `@abstractmethod` on these - that was in the original, no? (a little hard to find in GH web interface)

Reply:
nope. adding it breaks things.. messsaged in discord..